### ISSUE TITLE
Create a snapshot of a remote to assign it a DOI

### ISSUE THREAD

**Issue by pafonta** [iterative/dvc#5450]
Dear DVC team,

It would be great if one could **create a snapshot of a DVC remote**.

In academia, this makes sense to share a specific revision of data and models used in a scientific paper.

Using a live DVC remote (S3, HTTP, ...) isn't suitable here because one cannot assign a DOI ([Wikipedia](https://en.wikipedia.org/wiki/Digital_object_identifier)) to it.

The basic implementation could be a `dvc remote snapshot <remote name> ...` command.

However, one could argue that introducing snapshots could benefit to more use cases.

Then `dvc snapshot remote <remote name> ...` could then makes more sense.

The original discussion with @pared is [here](https://discuss.dvc.org/t/archive-share-a-snapshot-of-a-dvc-remote/664/3). 

Have a nice day!

**Comment by aerubanov** [comment]
Hi @pafonta !

I`m working on PR #4860 which add the support of OSF storage as a DVC remote. [OSF](https://help.osf.io/hc/en-us) allow to create project that combine files (from github for example) and data (dvc cache), freez it, assign DOI and publish. I think it may to solve this problem, at least partialy.

Unfortunately, work has been temporarily halted as the maintainers are now concentrating on preparing the dvc 2.0 release. I hope that we will continue to work after the release.

**Comment by efiop** [comment]
@pafonta Btw, do you only need the data for the current workspace? Kinda like with git you would create an archive from `master` branch, right? If so, it is the same as doing `dvc push` to a new remote, it will only push data that is used in the current workspace.

**Comment by pafonta** [comment]
Hi @aerubanov !

Thank you for having had a look at this feature request.

Your suggestion of using the coming integration between DVC and OSF makes sense.

However, OSF doesn't support **DOI versioning** yet ([source](https://help.osf.io/hc/en-us/articles/360019931013-Create-DOIs)).

DOI versioning lets one get a DOI for all the revisions of one's work but also a DOI for each new revisions of one's work.

To enable reproducible science, one would require DOI versioning. Indeed, a specific revision would need to be identifiable.

Besides, IMO, it would be good to keep the separation between the goal and the implementation as DVC does everywhere.

What I mean is that being able to assign a DOI from DVC wouldn't need to be associated to a specific service like OSF.

That's just a comment. I understood your suggestion was just to help and not to say that OSF is _the_ DOI provider to use.

**Comment by pafonta** [comment]
Hello @efiop !

That would indeed be great to be able to select if one wants to snapshot the entire history or just the HEAD.

I have actually a need for such feature ;) But more as a nice-to-have at the moment.

**Comment by pafonta** [comment]
Hello @aerubanov !

**TL;DR**
Another way of framing this feature is to do what GitHub does [here](https://guides.github.com/activities/citable-code/) but for DVC.

Maybe there are some contextual information worth reading there. I mentioned it in the description of the issue.

> The original discussion with @pared is [here](https://discuss.dvc.org/t/archive-share-a-snapshot-of-a-dvc-remote/664/3). 

So, yes, the snapshot would need to be an archive.

It doesn't need to be human-readable. Indeed, the goal is then to do a `dvc pull` in the `git` repository. It will make the content of the DVC remote human-readable. Besides, the directory structure will then be usable.

> What about files that were not added via dvc add, but are listed as dependencies?

Do you mean [external dependencies](https://dvc.org/doc/user-guide/external-dependencies)?

> If I understand correctly, the DOI assignment is possible after the publication of the object.

Yes.

> As I understand it, the user will publish the snapshot using some service that supports DOI assignment (Mendeley data, etc.). Is this true?

Yes.
